gobblin-iceberg/src/test/java/org/apache/gobblin/iceberg/writer/GobblinMCEWriterTest.java

```diff
@@ -0,0 +1,250 @@
+    addTableStatus(otherDb, "otherDatasetPath");
+
+    BiFunction<String, String, Void> verifyMocksCalled = new BiFunction<String, String, Void>(){
+      private int timesCalled = 0;
+      @Override
+      @SneakyThrows
+      public Void apply(String dbName, String tableName) {
+        timesCalled++;
+
+        // also validates that order is maintained since all writers after an exception should reset instead of write
+        Mockito.verify(mockWriter, Mockito.times(timesCalled)).writeEnvelope(
+            Mockito.any(RecordEnvelope.class), Mockito.anyMap(), Mockito.anyMap(), Mockito.any(HiveSpec.class));
+        Mockito.verify(exceptionWriter, Mockito.times(timesCalled)).writeEnvelope(
+            Mockito.any(RecordEnvelope.class), Mockito.anyMap(), Mockito.anyMap(), Mockito.any(HiveSpec.class));
+        Mockito.verify(exceptionWriter, Mockito.times(1)).reset(dbName, tableName);
+        Mockito.verify(mockWriter, Mockito.times(1)).reset(dbName, tableName);
+        return null;
+      }
+    };
+
+    writeToMetadataWriters(gmceBuilder.build());
+    verifyMocksCalled.apply(dbName, tableName);
+
+    // Another exception for same dataset but different db
+    when(mockTable.getDbName()).thenReturn(dbName2);
+    writeToMetadataWriters(gmceBuilder.build());
+    verifyMocksCalled.apply(dbName2, tableName);
+
+    // exception thrown because exceeds max number of datasets with errors
+    when(mockTable.getDbName()).thenReturn(otherDb);
+    Assert.expectThrows(IOException.class, () -> writeToMetadataWriters(gmceBuilder.setDatasetIdentifier(DatasetIdentifier.newBuilder()
+        .setDataPlatformUrn("urn:namespace:dataPlatform:hdfs")
+        .setNativeName("someOtherDB/testTable")
+        .build()).build()));
+  }
+
+  @Test(dataProvider = "AllowMockMetadataWriter")
+  public void testGetAllowedMetadataWriters(List<String> metadataWriters) {
+    Assert.assertNotEquals(mockWriter.getClass().getName(), exceptionWriter.getClass().getName());
+    gmceBuilder.setAllowedMetadataWriters(metadataWriters);
+    List<MetadataWriter> allowedWriters = GobblinMCEWriter.getAllowedMetadataWriters(
+        gmceBuilder.build(),
+        Arrays.asList(mockWriter, exceptionWriter));
+
+    Assert.assertEquals(allowedWriters.size(), 2);
+    Assert.assertEquals(allowedWriters.get(0).getClass().getName(), mockWriter.getClass().getName());
+    Assert.assertEquals(allowedWriters.get(1).getClass().getName(), exceptionWriter.getClass().getName());
+  }
+
+  @DataProvider(name="AllowMockMetadataWriter")
+  public Object[][] allowMockMetadataWriterParams() {
+    initMocks();
+    return new Object[][] {
+        {Arrays.asList(mockWriter.getClass().getName(), exceptionWriter.getClass().getName())},
+        {Collections.emptyList()}
+    };
+  }
+
+  private void initMocks() {
+    MockitoAnnotations.initMocks(this);
+    mockWriter = Mockito.mock(MetadataWriter.class);
+    exceptionWriter = Mockito.mock(HiveMetadataWriter.class);
```

Comment:
This is kind of a hack so I could have 2 `MetadataWriter` mocks with different class names